// @tsparticles/engine/umd/Utils/RgbColorManager.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "./NumberUtils.js"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.RgbColorManager = void 0;
    const NumberUtils_js_1 = require("./NumberUtils.js");
    class RgbColorManager {
        constructor() {
            this.key = "rgb";
            this.stringPrefix = "rgb";
        }
        handleColor(color) {
            const colorValue = color.value, rgbColor = colorValue.rgb ?? color.value;
            if (rgbColor.r !== undefined) {
                return rgbColor;
            }
        }
        handleRangeColor(color) {
            const colorValue = color.value, rgbColor = colorValue.rgb ?? color.value;
            if (rgbColor.r !== undefined) {
                return {
                    r: (0, NumberUtils_js_1.getRangeValue)(rgbColor.r),
                    g: (0, NumberUtils_js_1.getRangeValue)(rgbColor.g),
                    b: (0, NumberUtils_js_1.getRangeValue)(rgbColor.b),
                };
            }
        }
        parseString(input) {
            if (!input.startsWith(this.stringPrefix)) {
                return;
            }
            const regex = /rgba?\(\s*(\d+)\s*,\s*(\d+)\s*,\s*(\d+)\s*(,\s*([\d.%]+)\s*)?\)/i, result = regex.exec(input), radix = 10, minLength = 4, defaultAlpha = 1;
            return result
                ? {
                    a: result.length > minLength ? (0, NumberUtils_js_1.parseAlpha)(result[5]) : defaultAlpha,
                    b: parseInt(result[3], radix),
                    g: parseInt(result[2], radix),
                    r: parseInt(result[1], radix),
                }
                : undefined;
        }
    }
    exports.RgbColorManager = RgbColorManager;
});
