// @tsparticles/plugin-emitters/umd/EmittersPlugin.js
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "@tsparticles/engine", "./Options/Classes/Emitter.js"], factory);
    }
})(function (require, exports) {
    "use strict";
    var __syncRequire = typeof module === "object" && typeof module.exports === "object";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.EmittersPlugin = void 0;
    const engine_1 = require("@tsparticles/engine");
    const Emitter_js_1 = require("./Options/Classes/Emitter.js");
    class EmittersPlugin {
        constructor(engine) {
            this._engine = engine;
            this.id = "emitters";
        }
        async getPlugin(container) {
            const { Emitters } = await (__syncRequire ? Promise.resolve().then(() => __importStar(require("./Emitters.js"))) : new Promise((resolve_1, reject_1) => { require(["./Emitters.js"], resolve_1, reject_1); }).then(__importStar));
            return new Emitters(this._engine, container);
        }
        loadOptions(options, source) {
            if (!this.needsPlugin(options) && !this.needsPlugin(source)) {
                return;
            }
            if (source?.emitters) {
                options.emitters = (0, engine_1.executeOnSingleOrMultiple)(source.emitters, (emitter) => {
                    const tmp = new Emitter_js_1.Emitter();
                    tmp.load(emitter);
                    return tmp;
                });
            }
            const interactivityEmitters = source?.interactivity?.modes?.emitters;
            if (interactivityEmitters) {
                if ((0, engine_1.isArray)(interactivityEmitters)) {
                    options.interactivity.modes.emitters = {
                        random: {
                            count: 1,
                            enable: true,
                        },
                        value: interactivityEmitters.map((s) => {
                            const tmp = new Emitter_js_1.Emitter();
                            tmp.load(s);
                            return tmp;
                        }),
                    };
                }
                else {
                    const emitterMode = interactivityEmitters;
                    if (emitterMode.value !== undefined) {
                        const defaultCount = 1;
                        if ((0, engine_1.isArray)(emitterMode.value)) {
                            options.interactivity.modes.emitters = {
                                random: {
                                    count: emitterMode.random.count ?? defaultCount,
                                    enable: emitterMode.random.enable ?? false,
                                },
                                value: emitterMode.value.map((s) => {
                                    const tmp = new Emitter_js_1.Emitter();
                                    tmp.load(s);
                                    return tmp;
                                }),
                            };
                        }
                        else {
                            const tmp = new Emitter_js_1.Emitter();
                            tmp.load(emitterMode.value);
                            options.interactivity.modes.emitters = {
                                random: {
                                    count: emitterMode.random.count ?? defaultCount,
                                    enable: emitterMode.random.enable ?? false,
                                },
                                value: tmp,
                            };
                        }
                    }
                    else {
                        const emitterOptions = (options.interactivity.modes.emitters = {
                            random: {
                                count: 1,
                                enable: false,
                            },
                            value: new Emitter_js_1.Emitter(),
                        });
                        emitterOptions.value.load(interactivityEmitters);
                    }
                }
            }
        }
        needsPlugin(options) {
            if (!options) {
                return false;
            }
            const emitters = options.emitters;
            return (((0, engine_1.isArray)(emitters) && !!emitters.length) ||
                emitters !== undefined ||
                (!!options.interactivity?.events?.onClick?.mode &&
                    (0, engine_1.isInArray)("emitter", options.interactivity.events.onClick.mode)));
        }
    }
    exports.EmittersPlugin = EmittersPlugin;
});
